### ISSUE TITLE
[aot_autograd] wrong gradients for `torch.linalg.pinv` and `torch.linalg.matrix_sqrth` with complex input under torch.compile (aot_eager and inductor); eager, functorch and Dynamo-only are correct

### ISSUE THREAD

**Issue by cuiliaomei-beep** [pytorch/pytorch#197084]
### 🐛 Describe the bug

**Summary.** Eager and torch.compile do not produce the same gradient for `torch.linalg.pinv` and `torch.linalg.matrix_sqrth` on complex input: eager, `torch.func.grad` and `torch.compile(backend="eager")` agree with a complex128 reference to 1e-5, while `aot_eager` and `inductor` return a different tensor (max error 1.7e+01 on a (4, 3) complex64 input) and `gradcheck` fails on the compiled function. A user who compiles a model with a complex pseudo-inverse sees no error and no warning; training follows a wrong gradient, and real-valued inputs behave correctly, so the problem is confined to complex parameters. Because the divergence is introduced by how AOTAutograd traces the backward formula rather than by the user's program, every compiled backend that goes through AOTAutograd is affected in the same way.

The gradient of a loss that goes through `torch.linalg.pinv` of a **complex** matrix is wrong once the function is compiled. `gradcheck` passes in eager and fails on the compiled function.

```python
import torch

torch.manual_seed(0)


def loss(a):
    return (torch.linalg.pinv(a).abs() ** 2).sum()


def grad(fn, a):
    a = a.detach().clone().requires_grad_()
    fn(a).backward()
    return a.grad


a = torch.randn(4, 3, dtype=torch.complex64)
truth = grad(loss, a.to(torch.complex128))
for name, fn in [("eager", loss), ("compile(backend=eager)", torch.compile(loss, backend="eager")),
                 ("compile(aot_eager)", torch.compile(loss, backend="aot_eager")), ("compile(inductor)", torch.compile(loss))]:
    torch._dynamo.reset()
    print(f"{name:<24} max |grad - truth| = {(grad(fn, a).to(torch.complex128) - truth).abs().max().item():.1e}")

a128 = torch.randn(4, 3, dtype=torch.complex128, requires_grad=True)
print("gradcheck eager   :", torch.autograd.gradcheck(loss, (a128,), eps=1e-6, atol=1e-4, raise_exception=False))
print("gradcheck compiled:", torch.autograd.gradcheck(torch.compile(loss, backend="aot_eager"), (a128,), eps=1e-6, atol=1e-4, raise_exception=False))
```

Output (torch 2.14.0+cpu; same on the 2026-09-11 nightly):

```
eager                    max |grad - truth| = 1.3e-05
compile(backend=eager)   max |grad - truth| = 1.3e-05
compile(aot_eager)       max |grad - truth| = 1.7e+01
compile(inductor)        max |grad - truth| = 1.7e+01
gradcheck eager   : True
gradcheck compiled: False
```

The compiled gradient is not slightly off, it is a different tensor (first row eager `[-0.5897+0.0005j, -0.5655-0.2723j, -0.5190-0.7107j]`, compiled `[-0.0124+0.0438j, -1.5919-0.0432j, -1.3720-2.4505j]`). With a `float32` input all four agree (error 2.5e-5 for every path).

Localisation:

- `torch.compile(backend="eager")` (Dynamo only, no AOTAutograd) is correct; `torch.func.grad(loss)` and `torch.func.functionalize(loss)` are correct; `torch._functorch.aot_autograd.aot_function(loss, nop, nop)` without Dynamo is already wrong → the problem is in how AOTAutograd traces the backward of `linalg_pinv`.
- The traced forward keeps `aten.linalg_pinv.atol_rtol_tensor` as one node; the traced backward consists of `_conj`, `mm`, `transpose`, `mul`, `sub`, `neg`, `sgn`, `pow`, `expand`, `clone`, `detach` — i.e. the C++ `pinv_backward` formula (which uses `.mH`) traced into aten ops. The `aten._conj.default` nodes are the suspicious part: the conjugation appears to be lost or applied to the wrong operand somewhere between tracing and execution.
- Other complex linear-algebra gradients traced the same way are right: `pinv(hermitian=True)`, `inv`, `solve`, `svd`/`svdvals`, `qr`, `eigvalsh`, `lstsq`, `det`, `cholesky`, `matrix_exp`, `matrix_norm` all match eager under `aot_eager` and `inductor`.

A second operator with the same signature: `torch.linalg.matrix_sqrth`. For `loss(a) = (matrix_sqrth(a @ a.mH + I).abs() ** 2).sum()` with a `complex64` (4, 4) input, eager / Dynamo-only error vs complex128 is 2.2e-6, `aot_eager` and `inductor` give 9.6e-1; float32 agrees everywhere. Its traced backward again contains `aten._conj.default` (plus `_linalg_eigh`, `mm`, `div`, ...). Same numbers on Linux / Python 3.11 (2.14.0 and the 2026-09-12 nightly). A sweep over every OpInfo operator that supports complex64 autograd (3 samples each, eager vs `aot_eager`) flags exactly `linalg.pinv`, `pinverse` and `linalg.matrix_sqrth` — nothing else — so this is not a general complex-autograd problem but something specific to how these two backward formulas are traced.

Expected: the compiled gradient equals the eager gradient (and passes gradcheck), as it does for real inputs.

### Error logs

No error; silent wrong gradient (gradcheck output above).

### Versions

Reproduced on:
- Windows 11, Python 3.14.7, torch 2.14.0+cpu (MSVC 19.44 for the Inductor runs)
- the same machine with the nightly wheel torch 2.15.0.dev20260911+cpu
- Linux (Kaggle image), Python 3.11, torch 2.14.0+cpu and torch 2.15.0.dev20260912+cpu

CPU only; the divergence is in tracing, so the device should not matter.

```
PyTorch version: 2.14.0+cpu
Is debug build: False
CUDA used to build PyTorch: None
OS: Microsoft Windows 11 Pro for Workstations (10.0.26200 64-bit)
Python version: 3.14.7 (64-bit runtime)
Is CUDA available: False
CPU: 11th Gen Intel(R) Core(TM) i9-11900T @ 1.50GHz
[pip3] numpy==2.5.2
[pip3] torch==2.14.0+cpu
[pip3] torchvision==0.29.0+cpu
```

cc @ezyang @anjali411 @dylanbespalko @mruberry @nikitaved @amjames @jianyuh @walterddr @Lezcano @chauhang @penguinwu @bdhirsh @bobrenjc93 @aorenste

**Comment by cuiliaomei-beep** [comment]
Correction on the cause: it is not the traced backward formula. The **forward** value of complex `torch.linalg.pinv` is already wrong whenever a `TorchDispatchMode` is active, and `torch.compile`'s runtime wrapper runs every graph under one (`_AnalyzeCustomOpInputOutputMode`). Details and the affected set (`linalg.pinv` / `pinverse`, `linalg.polar`, `linalg.matrix_sqrth`) are in #198118; here is the gradient side of it, all in eager:

```python
import torch
from torch.utils._python_dispatch import TorchDispatchMode

class Forward(TorchDispatchMode):            # only forwards
    def __torch_dispatch__(self, func, types, args=(), kwargs=None):
        return func(*args, **(kwargs or {}))

torch.manual_seed(0)
def loss(a):
    return (torch.linalg.pinv(a).abs() ** 2).sum()

def grad(fn, a, mode=None):
    a = a.detach().clone().requires_grad_()
    if mode is None:
        fn(a).backward()
    else:
        with mode:
            fn(a).backward()
    return a.grad

a = torch.randn(4, 3, dtype=torch.complex64)
truth = grad(loss, a.to(torch.complex128))
g_eager, g_mode = grad(loss, a), grad(loss, a, Forward())
torch._dynamo.reset()
g_compiled = grad(torch.compile(loss, backend="aot_eager"), a)
print("eager          |grad - truth|", (g_eager.to(torch.complex128) - truth).abs().max().item())
print("under mode     |grad - truth|", (g_mode.to(torch.complex128) - truth).abs().max().item(), "   |mode - compiled|", (g_mode - g_compiled).abs().max().item())
a2 = a.detach().clone().requires_grad_()
with Forward():
    l = loss(a2)                               # forward under the mode, backward outside it
l.backward()
print("fwd in mode, bwd outside |grad - truth|", (a2.grad.to(torch.complex128) - truth).abs().max().item())
```

```
eager          |grad - truth| 1.2571577899445539e-05
under mode     |grad - truth| 17.399362363508732    |mode - compiled| 0.0
fwd in mode, bwd outside |grad - truth| 17.399362363508732
```

The eager gradient computed under a forwarding mode is bit-identical to the compiled one, and it stays wrong when only the forward runs under the mode, so the backward formula is fine and the saved forward output is what is wrong (`pinv` comes back as `vh.mT @ diag(1/s) @ u.mH`, one internal conjugation dropped). `matrix_sqrth` shows the same forward difference under the mode. So the fix belongs to #198118 (the kernels' internal `.mH()` under the `Python` dispatch key), not to AOTAutograd; the labels `module: aotdispatch` / `module: pt2-dispatcher` here are misleading, `module: complex` / `module: linear algebra` / `module: __torch_dispatch__` fit better.
